src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -562,7 +562,22 @@ void CoreWorkerDirectTaskSubmitter::RequestNewWorkerIfNeeded(
             }
           } else if (reply.canceled()) {
             RAY_LOG(DEBUG) << "Lease canceled " << task_id;
-            RequestNewWorkerIfNeeded(scheduling_key);
+            if (reply.placement_group_removed()) {
```

Comment:
@rkooo567 , The reason why it doesn't throw TaskCanceledError before is that we marded the reply canceled in server raylet, but when request client received the reply, it will continue to invoke RequestNewWorkerIfNeeded to retry for some reason. so I add a `placement_group_removed` flag to the reply.

Reply:
Prefer enum like CancelType rather than placement_group_removed (cancel type has pg removed and task cancelled)

Reply:
got it, good idea!